text/0240-app-telemetry-schema-vision-roadmap.md

```diff
@@ -0,0 +1,659 @@
+Telemetry Schema.
+* `gen-client sdk`: Generates a Client SDK from a Component or Resolved
+Telemetry Schema.
+* `gen-client api`: Produces a Client API from a Component or Resolved
+Telemetry Schema.
+
+A Plugin System is planned to allow community contributions to the OTel Weaver
+tool. Proposed plugins include:
+
+* An example of plugin for gathering resolved telemetry schemas from external
+dependencies.
+* An example of compatibility checker plugin to ensure successive versions of
+the same Telemetry Schema follow specified compatibility rules.
+
+Additional plugins and tools are anticipated to be developed by the broader
+community, leveraging the OTel Weaver Plugin System and the standardized format
+of the Resolved Telemetry Schema.
+
+## Links
+
+- [Positional Paper: Schema-First Application Telemetry](https://research.facebook.com/publications/positional-paper-schema-first-application-telemetry/)
+- [A benchmark to understand the role of knowledge graphs on Large Language Model's accuracy for question answering on enterprise sql databases](https://arxiv.org/pdf/2311.07509.pdf#:~:text=The%20results%20of%20the%20benchmark%20provide%20evidence%20that%20supports%20our,LLM%20without%20a%20Knowledge%20Graph)
+- [MetricFlow framework](https://docs.getdbt.com/docs/build/about-metricflow)
+
+## Appendices
+
+### Appendix A - Component Telemetry Schema
+
+The structure and format of the component telemetry schema is given here as an
+example and corresponds to the author's vision (not yet validated) of what the
+structure/format of a component telemetry schema could be. The definitive
+structure and format of this component schema will be discussed and finalized
+later in a dedicated OTEP.
+
+```yaml
+# This file describes the structure and format of a component telemetry schema.
+# A component telemetry schema can be used for three types of purposes:
+# - defining the telemetry schema for an application or a service
+# - defining the telemetry schema for a library
+# - defining a semantic convention registry
+file_format: 1.2.0
+schema_url: <url-of-the-current-component-schema>
+# This optional field allows to specify the parent schema of the current schema.
+# The parent schema is a resolved schema.
+parent_schema_url: <url-of-a-parent-resolved-schema>
+
+# This optional section allows for importing a semantic convention registry
+# from a git repository containing a set of semantic convention files. It is
+# also possible to import file by file.
+semantic_conventions:
+  - git_url: <git-url-of-the-semantic-conventions-repository>
+    path: <path-to-the-semantic-conventions-directory-inside-the-git-repo>
+  - url: <url-of-the-semantic-conventions-file>
+
+# The resource field is defined when the component schema is that of an
+# application (as opposed to that of a library). The resource field contains a
+# list of local references to attributes defined in the shared catalog within
+# this file.
+resource:
```

Comment:
This is interesting. When I think of a "component schema" I think of a compile unit, i.e. the schema of the telemetry is "fixed" after the code is compiled. But the resource attributes are not scoped to source code, they can be quite often added dynamically depending on where that compiled unit is deployed. Strictly speaking, you can even say that about other attributes, e.g. a processor in the collector could add certain attributes to telemetry (e.g. "processing timestamp"). So the resulting schema seen at rest may not match the component schema. 

How are the dynamic enrichments should be accounted for in the schema management lifecycle?

Reply:
For this specific case, I was thinking of offering the possibility to define the value of an attribute dynamically by referencing an environment variable (or any other value injection mechanism at the time of deployment). Thus, the attribute is defined in the schema with a clearly identified placeholder. The value of this placeholder is resolved at the time of deployment or by the SDK at runtime in a particular environment. We can also support cases where the identifiers of the attributes themself are not known before deployment, but in this case, these attributes will not be controlled/known at the time of the build.

Reply:
but I am not talking about values, rather the actual attributes. If your app is deployed on AWS, it may get AWS-specific resource attributes, which it wouldn't get on GCP.

Reply:
Yes, I had a doubt, which is why the second part of my response includes the case you are talking about, that is, the case where certain attributes (identifier and value) are only known at the time of deployment.

Reply:
seems like it's often going to be the case, so if the ingestion pipeline does any kind of schema validation it would need to know these deployment-enriched materialized schemas